waiter/src/waiter/scheduler/cook.clj

```diff
@@ -0,0 +1,523 @@
+          (catch ExceptionInfo ex
+            (throw ex))
+          (catch Exception ex
+            (throw (ex-info "cook-backend-port metadata must parse to a positive integer"
+                            {:cook-backend-port cook-backend-port} ex)))))
+      {:jobs [(cond-> {:application {:name name
+                                     :version version}
+                       :command cmd
+                       :cpus cpus
+                       :disable-mea-culpa-retries true
+                       :env environment
+                       :executor "cook"
+                       :labels (cond-> {:backend-proto backend-proto
+                                        :health-check-url health-check-url
+                                        :service-id service-id
+                                        :source "waiter"
+                                        :user run-as-user}
+                                 cook-backend-port (assoc :backend-port cook-backend-port))
+                       :max-retries 1
+                       ;; extend max runtime past instance expiry as there may be a delay in killing it
+                       :max-runtime (-> instance-expiry-mins
+                                        (+ instance-expiry-adjustment-mins)
+                                        t/minutes
+                                        t/in-millis)
+                       :mem mem
+                       :name (str service-id "." job-uuid)
+                       :ports ports
+                       :priority instance-priority
+                       :uuid job-uuid}
+                (and docker-image-namespace docker-image-name docker-image-label)
+                (assoc :container {:docker {:force-pull-image false
+                                            :image (str "namespace:" docker-image-namespace ","
+                                                        "name:" docker-image-name ","
+                                                        "label:" docker-image-label)
+                                            :network "HOST"}
+                                   :type "docker"}))]})))
+
+(defn create-job
+  "Create and start a new Cook job specified by the service-description."
+  [cook-api service-id service-description service-id->password-fn home-path-prefix instance-priority]
+  (log/info "creating a new job for" service-id)
+  (->> (create-job-description service-id service-description service-id->password-fn home-path-prefix instance-priority)
+       (post-jobs cook-api)))
+
+(defn determine-instance-priority
+  "Determines the instance priority based on allowed (sorted by highest first) priorities and unavailable priorities."
+  [allowed-priorities unavailable-priorities]
+  (or (some #(and (not (contains? unavailable-priorities %)) %) allowed-priorities)
+      (last allowed-priorities)))
+
+(defn launch-jobs
+  "Launch num-instances jobs specified by the service-description."
```

Comment:
A comment would be useful describing how priorities are assigned.

Reply:
Added comment.